media/js/base/mozilla-client.js

```diff
@@ -158,16 +158,32 @@ if (typeof Mozilla === 'undefined') {
 
         for (var i = 0; i < versions.length; i++) {
             var latestVerArr = versions[i].split('.');
+            var userVerNumber = 0;
+            var latestVerNumber = 0;
+
+            // Make sure both latest and user array lengths match
+            while (latestVerArr.length < userVerArr.length) {
+                latestVerArr.push('0');
+            }
+            while (userVerArr.length < latestVerArr.length) {
+                userVerArr.push('0');
+            }
 
             // Only check the major version in non-strict comparison mode
             if (!strict) {
                 latestVerArr.length = 1;
             }
 
-            for (var j = 0; j < userVerArr.length; j++) {
-                if (Number(userVerArr[j]) < Number(latestVerArr[j] || 0)) {
+            for (var j = 0; j < latestVerArr.length; j++) {
+                userVerNumber = Number(userVerArr[j]);
+                latestVerNumber = Number(latestVerArr[j]);
+
+                if (userVerNumber < latestVerNumber) {
                     isUpToDate = false;
                     break;
+                } if (userVerNumber > latestVerNumber) {
+                    isUpToDate = true;
+                    break;
```

Comment:
The new logic here is sound as far as I can tell/test. However, it's really hard to follow this function - the complexity has exceeded reasonable variable names. If you have the time (and freshness of mind), it would be great to add some detailed comments.